core/domain/exp_domain.py

```diff
@@ -959,6 +963,86 @@ def validate(self, strict=False):
+                        raise utils.ValidationError(
+                            'Expected card_is_checkpoint of end state to be '
+                            'False but found it to be %s'
+                            % self.states[state_name].card_is_checkpoint
+                        )
+
+        # Check if checkpoint count is between 1 and 8, inclusive.
+        checkpoint_count = 0
+        for state_name, state in self.states.items():
+            if state.card_is_checkpoint:
+                checkpoint_count = checkpoint_count + 1
+        if not 1 <= checkpoint_count <= 8:
+            raise utils.ValidationError(
+                'Expected checkpoint count to be atleast 1 and atmax 8 '
+                'but found it to be %s'
+                % checkpoint_count
+            )
+
+        # Check if a state marked as checkpoint is bypassable.
+        end_state_names = []
+        for state_name, state in self.states.items():
+            if state.interaction.id == 'EndExploration':
+                end_state_names.append(state_name)
+
+        # For every state which is marked as a checkpoint and is not the initial
+        # state we remove it from the states dict. Then we make a list of
+        # unseen states. If all end states are unseen on removing a state,
+        # this means that after removing the state it is impossible to reach
+        # any end states, thus the user cannot bypass that state. However, if
+        # the unseen states list contains at least one end state, then the state
+        # is bypassable.
+        for state_name, state in self.states.items():
+            if state_name == self.init_state_name:
+                continue
+            elif state.card_is_checkpoint:
+                new_states = copy.deepcopy(self.states)
+                new_states.pop(state_name)
+                processed_queue = []
+                curr_queue = [self.init_state_name]
+
+                while curr_queue:
+                    if curr_queue[0] == state_name:
+                        curr_queue.pop(0)
+                        continue
+                    curr_state_name = curr_queue[0]
+                    curr_queue = curr_queue[1:]
+                    if not curr_state_name in processed_queue:
+                        processed_queue.append(curr_state_name)
+
+                        curr_state = new_states[curr_state_name]
+
+                        if not curr_state.interaction.is_terminal:
+                            all_outcomes = (
+                                curr_state.interaction.get_all_outcomes())
+                            for outcome in all_outcomes:
+                                dest_state = outcome.dest
+                                if (dest_state not in curr_queue and
+                                        dest_state not in processed_queue):
+                                    curr_queue.append(dest_state)
+                unseen_states = list(
```

Comment:
If you set a flag and exit early as soon as you discover an end state, then you don't need to do this calculation. You can just look at the flag.

Reply:
I think we don't even need a flag. We can raise a ValidationError as soon as we discover an end state. Does that sound good to you?

Reply:
Hm, either approach could work, but I think a clearly-named flag (and raising later) would probably be better for code maintainability.

Reply:
Done